### ISSUE TITLE
[feature request] support a paste operator in join.as() syntax

### ISSUE THREAD

**Issue by jonseymour** [influxdata/kapacitor#435]
As it currently stands, the field names in output stream if a Kapacitor join node are assigned prefixes of the form \<aliasN\>. where \<aliasN> is the alias assigned to the Nth input stream by the as() property method of the join node.

So, for example:

```
var metric1 = stream
        |from()
                .database('sampledb')
                .retentionPolicy('default')
                .measurement('sample')
                .groupBy('node')
        |window()
                .align()
                .period(5m)
                .every(5m)
        |shift(-5m)
        |mean('metric1').as('metric1')


var metric2 = stream
        |from()
                .database('sampledb')
                .retentionPolicy('default')
                .measurement('sample')
                .groupBy('node')
        |window()
                .align()
                .period(5m)
                .every(5m)
        |shift(-5m)
        |mean('metric2').as('metric2')

metric1
        |join(metric2)
           .as('m1', 'm2')
        |log()
```

will result in a stream with fields called 'm1.metric1' and 'm2.metric2'

It would be nice if the mandatory period in the resulting name could be avoided. 

The suggestion to avoid this is to support use of a trailing paste operator in the arguments to as()

So:

```
metric1
        |join(metric2)
           .as('s_#', 't_#')
```

would result in a stream with fields called s_metric1 and t_metric2 and

and:

```
metric1
        |join(metric2)
           .as('#', '#')
```

would result in a stream with fields called: metric1 and metric2.

'#' is chosen to mimic similar behaviour in the C pre-processor.

Note: A specification of as('foo##', 'foo##') would result in names called 'foo#metric1', 'foo#metric2'.

A specification of as('foo#bar', 'foo#bar') would result in names called foo#bar.metric1 and foo#bar.metric2.

**Comment by jonseymour** [comment]
An assertNotNill('metric1', 'metric4', ... etc) node might be useful for scripts that need to assert they are written with an assumption that the specified fields never contain nil

**Comment by nathanielc** [comment]
@jonseymour  I like the suggestion but the names of the joining nodes are not always available. If they were I would make 'as' optional. 

Example:

```javascript
var metric1 = stream
        |from()
                .database('sampledb')
                .retentionPolicy('default')
                .measurement('sample')
                .groupBy('node')
        |window()
                .align()
                .period(5m)
                .every(5m)
        |shift(-5m)
        |mean('metric1').as('metric1')

var metric2 = stream
        |from()
                .database('sampledb')
                .retentionPolicy('default')
                .measurement('sample')
                .groupBy('node')
        |window()
                .align()
                .period(5m)
                .every(5m)
        |shift(-5m)
        |mean('metric2').as('metric2')
        |join(metric1)
           .as('m2', 'm1')
        |log()
```

While convention says that the above script is not readable and so its not used as much its still valid.

Maybe we can make try and populate the `as` prefixes with the var names be default. If they are not available then `as` would be required.

Also the `.` isn't all that important it just made sense to use something to delimit the prefix from the original name.

**Comment by jonseymour** [comment]
The paste operator would be optional of course - if you didn't specify it then current behaviour would apply.

The paste operator would, of course, provide an easy way for the user choose their own delimiter, as in the example above where I used s_# and t_#.
 
I also like your idea of making as() optional where possible.

**Comment by cnelissen** [comment]
+1 for making as() optional. IMO it should only be required if two nodes have fields with the same name, much like when you attempt a join in MySQL on two tables that share columns you get an error like: "column 'id' is ambiguous".

**Comment by jonseymour** [comment]
Closed via #698
